### ISSUE TITLE
When importing in python, pretty printing doesn't work

### ISSUE THREAD

**Issue by drmarkwslater** [ganga-devs/ganga#146]
On the little ganga branch, I can now import everything, submit jobs, etc. However, the pretty printing just doesn't work. I'm guessing this is because it's done in an IPython way with `_repr_pretty_` but if you call `__str__()` then that (sort of) works in vanilla python. What would be the best way to get this working in normal python?

**Comment by rob-c** [comment]
@drmarkwslater Does this work? http://stackoverflow.com/questions/17248383/pretty-print-by-default-in-python-repl

**Comment by milliams** [comment]
I guess it really comes down to: do we want to always do pretty printing? If so then we should just make the `__str__` method return what we want.

**Comment by drmarkwslater** [comment]
@rob-c Not sure I want to do something like that on import as it would change it for everything in the session...

@milliams That's the big question :) I personally couldn't think of a reason NOT to want it to print something sensible rather than <<Ganga Proxy Object blah blah>> but I may have missed something. This way it would only be restricted to the Ganga (proxy) objects as well.

Is this a solved problem in other API/library style packages? I don't know what the generally accepted solution is...

**Comment by milliams** [comment]
The relevant documentation for this is [`__repr__`](https://docs.python.org/3/reference/datamodel.html?highlight=__repr__#object.__repr__) and [`__str__`](https://docs.python.org/3/reference/datamodel.html?highlight=__repr__#object.__str__). The gist is that the `repr()` of an object should return a string which, if passed to `eval()` gives back an object just like the original. In this sense our current pretty printing is almost there (and is how `export`/`load` work). Calling `str()` on an object should return the "the 'informal' or nicely printable string representation of an object" which I guess could be identical to `repr()`, at least for proxies.

The two obvious caveats to this:
* new lines in the `repr()` output might mess things up so extra care there is needed
* objects like `jobs` should have a different `__str__` to their `_repr__` since their primary purpose is printing. Otherwise their `__repr__` can be a `<... object at ..>`-type string.

I think this should be done on the proxy layer where possible since that is where doing `eval(repr(j))` will actually work due to the proxy constructors taking schema items. However, the heavy lifting should be done behind the proxy to stop the proxy layer getting too thick.

**Comment by rob-c** [comment]
I suppose a secondary question is do we want to change the printer when doing 'import ganga' within vanilla Python or change the printer within IPython.
Having 2 different ways of printing and keeping them in sync adds difficulties which we may not (or may) want to face.
@drmarkwslater I would have thought it's a reasonable expectation that the printer may be changed by importing something such as ganga. It should be possible to add a configuration option to disable/enable the setting of the printer but I can't think of many people not preferring or wanting to use pprint. But that's just my opinion.

**Comment by drmarkwslater** [comment]
@rob-c If that's OK with everyone then I'm happy with that as well - my knee-jerk reaction is always for `import ganga` to do the minimal of messing :)

From what @milliams said, I think I'd lean towards changing the more general `__str__` and `__repr__` to behave as expected then you wouldn't need a separate IPtyhon version.  Though maybe I've misunderstood :)

Going one step further - do we still need all that `VPrinter` stuff? I feel that can be retired (or at least simplified) as well...

**Comment by milliams** [comment]
I think the simplest solution would be to change `__str__` to do the sensible thing. We can then get rid of the IPython hooks for displaying since they won't be doing anything extra.

`VPrinter` is useful in a way as it provides a way of generating the string based on a GangaObject. It can perhaps be simplified though.

**Comment by rob-c** [comment]
@drmarkwslater I think you're correct that's probably for the best.

@milliams Wouldn't it be worth just pointing ```__str__``` (and maybe ```__repr```) to the pretty repr used by IPython. I think it will always look for these other functions first if they exist (I could be wrong). We could change the IPython printer but then I wonder if this is guaranteed to be the same for all (reasonable) versions of IPython we could support.

**Comment by egede** [comment]
Don't forget that we have implemented the compressed printout for long lists. So a list of 1000 files in an LHCbDataset just prints out a summary rather than the 1000 entries. See below. This behaviour should be preserved.

```python
Ganga In [11]: d = LHCbDataset([LocalFile(str(i)) for i in range(1000)])

[22:11:31]
Ganga In [12]: print d
 LHCbDataset (
   depth = 0,
   treat_as_inputfiles = False,
   persistency = None,
   files = [1000 Entries of type 'LocalFile'] ,
   XMLCatalogueSlice =    LocalFile (
     namePattern =    '',
     compressed = False,
     localDir =    ''
   ) 
 )
```

**Comment by alexanderrichards** [comment]
hi all, I came to this discussion late. From my point of view, I've never been a fan of the VPrinter so would be happy to see it go. I would argue for having just python standard `__str__` and `__repr__`. As @milliams rightly says the convention is that `__repr__` should return an eval'able string while `__str__` can return whatever. In this case since `__repr__` is called when python displays return values or simply someone does:
```python
>>> j=Job()
>>> j
```
Then I think that our `__repr__` should be pretty printed and then `__str__` should match. I checked and newlines seem ok in eval (at least in my non-exhaustive test). The problem will come with the list contractions mentioned by @egede.

@milliams What did you envisage being done in the proxy layer vs the object layer? I would assume that the printing of proxies must be different as hidden schema attributes and non-exported methods would need hiding but cant think of much else that would be different.

**Comment by milliams** [comment]
This is my view on the difference between the three:
* `__repr__` should be fully `eval`able and should print the full contexts of the list. It should map to what the current `export()` does using the VPrinter.
* `_repr_pretty_` will always be called by IPython (even without manual hooks) instead of `repr()` when writing, e.g., `j` and so we needn't worry about `__repr__` being overly verbose. This pretty version can perform the contractions Ulrik mentions.
* `__str__` should not necessarily be `eval`able and so we can perform the contractions.

On the topic of their contents:
* `__repr__` probably shouldn't print non-copyable attributes since they will not be used when `eval`ing.
* `_repr_pretty_` and `__str__` can print them since you'll want to see the status if a job for example.

In summary:

Function: | `__repr__` | `_pretty_repr_` | `__str__` 
-------------|--------------|---------------------|----------------
Expand long lists | ✓ | ✗ | ✗ 
Show copyable | ✗ | ✓ | ✓

(any other differences and I'll update the table)

@alexanderrichards Perhaps nothing needs doing different in the proxy layer except as you say, hidden attributes.

**Comment by alexanderrichards** [comment]
Excellent use of the table feature ! :smile:

I am in complete agreement with all of this. In addition to the hidden attributes, the proxy must also, in a similar way to the non/copyable attributes, hide the protected attributes in the `__repr__` as these are essentially not settable to the proxy object.

@milliams can you update the table with two sections for the rows. One for objects and one for proxies. Then I think that this table should make it to our wiki/doc

**Comment by milliams** [comment]
I've updated the table. It's two separate tables since Markdown is quite restrictive. I've also added a row for `transient` since that is special too. I guess they shouldn't be shown in the `__repr__` but should in the others?

**Comment by alexanderrichards** [comment]
I think in the table, as we should use it in the doc, change:
copyable -> non-copyable
as I think it's more obvious what you mean then.

I'm trying to think of a common attribute that we have that is transient to see how that should behave. Can you think of any examples

**Comment by alexanderrichards** [comment]
also in the table it should be __repr_pretty__ rather than the other way around.

**Comment by milliams** [comment]
`Job.outputdir` is transient. It's something which is computed based on other attributes. Think of them like a:
```python
@property
def outputdir(self):
    return os.path.join('something', some_func(), 'etc')
```
without an associated setter. I guess you could think of them as `const` (even on the `GangaObject` layer).

**Comment by rob-c** [comment]
@alexanderrichards I'm a little hesitant to call for the replacement of VPrinter as it does a relatively good job of in a 'standard' way serializing an object and makes use of the same hooks as the XML code.
And it works. Plus we still need to address #82.

@milliams Is the table expected or actual behaviour?
Currently we use ```__repr__``` to generate the ```< Proxy object at 0x....>``` type responses from proxied objects atm. Something which I don't think is of a major benefit to anyone other than to remind us that the object is proxied.

Would we rather change our code so that our objects work with the default printers in both Python and IPython or do we prefer to change the (I)Python environments to suit our way of working?
(It's possible to modify the printer in both environments to point both to any of the above functions)

**Comment by milliams** [comment]
@rob-c The tables show our expected and desired behaviour. I think I'd prefer to change our way of working so that right things happen automatically in Python and IPython.

**Comment by rob-c** [comment]
OK, I'm more than happy to have ```__repr__``` do something a bit more sane :). I'm assuming people using ```import ganga``` are likely experienced users who don't need the 'hand-holding' provided by ```__repr_pretty__```.

I'm a little concerned that none of the above behind the proxy are suitable for replacing a lot of the logic in ```export(myJob)```.
I think modifying the ```__repr__``` to show non-copyable objects behind the proxy is more correct as the proxy is the thing which stops us from changing/copying/setting objects. If we want the exact representation of the object we want to know what this is set to.
You could argue this would allow ```import(myJobFile.txt)``` to subvert the proxy but it already should if it's working correctly imo.

I would propose:

GangaObject         | `__repr__` | `_repr_pretty_` | `__str__` 
------------------|------------|-----------------|----------------
Show transient | ✗ | ✗ | ✗

Proxy         | `__repr__` | `_repr_pretty_` | `__str__` 
------------------|------------|-----------------|----------------
Show transient | ✗ | ✓ | ✓

I can't think of a good reason to expose transients to behind the scenes strings if we expect ```__repr__``` to be ```eval```-able but it would be nice to see these values when in an interactive mode.

**Comment by drmarkwslater** [comment]
@rob-c Just a quick point - we're hoping that `import ganga` will basically replace the bootstrap in the medium-to-long term and I suspect that as soon as we start offering this, it might get quite a large take up  as it's a lot less hassle to just add this to your existing python scripts. I personally think having `import ganga` give a reliable and sensible environment irrespective of (I)Python is the way to go. 

Other than I'm happy with whatever conclusions are reached - I honestly didn't realise this was such a big issue :)

**Comment by alexanderrichards** [comment]
Wow ok a lot to catch up on since I went to lunch.

@rob-c 
> I'm a little hesitant to call for the replacement of VPrinter as it does a relatively good job of in a 'standard' way serializing an object and makes use of the same hooks as the XML code.
> And it works. Plus we still need to address #82.

These aren't to me good enough reasons to keep it around when standard python `__repr__` and `__str__` can do the same job. I think we should strive for simplicity and conciseness and it will make our code better and I don't think the answer is in the VPrinter.  As for #82 I would have to read it more carefully by I would be surprised if that too can't be solved in a nice way without the VPrinter.



@milliams @rob-c 
> I think I'd prefer to change our way of working so that right things happen automatically in Python and IPython.

I agree with this, especially as users who will be using the new `import ganga` will not have our lovely constructed environment and we should strive to give as consistent an experience as possible.


As for the transient issue, I also agree that we should not have them for `__repr__` as we wouldn't want to eval them. I'm afraid I didn't follow your reason @rob-c for not having __repr_pretty__ for objects print transients but having them in proxy. I'll re-read your post

**Comment by milliams** [comment]
@alexanderrichards I think using `VPrinter` as the mechanism for constructing the string for `__repr__` and `__str__` isn't a bad idea. It's code that does what we want and it works. Now, it might need changing to be able to support the different configurations mentioned above but it's not *bad* code. Unless there's a good reason to replace it?

**Comment by alexanderrichards** [comment]
@milliams I'll look again at the VPrinter code then but as I recall it's an implementation of the visitor pattern which I think is over the top for our needs so I'm afraid I disagree.

**Comment by milliams** [comment]
@alexanderrichards I agree that we should get rid of the visitor pattern if we can as it's unnecessarily complicated but it's easy enough to retrofit VPrinter (and VStreamer) to not use `accept()` etc.

**Comment by alexanderrichards** [comment]
well at least we agree on the need to remove the visitor pattern :wink: I think that we may have to keep in around in some form while we have all the `Node` nonsense. However I would like to also get rid of that (Another discussion maybe). I think though that if it's possible to simplify it to put it in the

**Comment by rob-c** [comment]
I think removing VPrinter is a nice direction to go in, but is somehow a mid to long term solution. A top down inspection of objects based upon the Schema is much clearer to understand but is a luxury.
At the moment it works as expected and produces sane results.

@milliams I somehow feel it would be better to start again if we want to remove the visitor pattern as we can make some sensible decisions about what to do rather than retro-fitting working code. But thats just my opinion.

@alexanderrichards Changing any of the code related to this will cause all prepared jobs to be un-prepared and re-prepared for users so I'd rather not make this change without being well aware we're braking things in a certain release version.
(The prepared app hash is calculated from this representation string so is whitespace sensitive and excessively fragile)

I'm under the impression that what users get from ```import ganga``` is access to 'everything' in the GPI (with some basic amount of the repos being initialized things such as this). And obviously they don't then launch an IPython session.
I think if we plan to keep support for IPython and add support for ```import ganga``` the only difference is where the default printer for each environment will look for the string representation of objects?

**Comment by alexanderrichards** [comment]
We have somewhat hijacked Marks issue here :smile:  I agree that what I am proposing is a long term goal but I thought that was the point of the discussion. I realise that we will have to get there in incremental steps which will very likely involve using the VPrinter as is and stripping the visitor pattern if possibly. Lets discuss this further in the meeting tomorrow.

**Comment by rob-c** [comment]
@drmarkwslater Are we able to close this or do we need to do something here in the Proxy to make sure that the correct string representation is called for proxied objects?

**Comment by drmarkwslater** [comment]
I'm afraid not:
```
>>> from ganga import jobs
>>> jobs
<Ganga.GPIDev.Lib.Registry.JobRegistry.JobRegistrySliceProxy object at 0xe92d50>
>>> jobs(0)
Job#0
>>> 
```
